### PR TITLE
Implement Role creation/validation for ops-user

### LINKED ISSUE #6504: Implementation of create/validate/repair of vSphere roles for ops-user
**User Statement:**
When the VCH is created the VIC should check that the appropriate roles for the ops-user are in place in vCenter. If not it should either create or repair the roles.
**Details:**
This issue should cover:
- APIs to create/validate/repair vSphere roles for ops-user
- Unit test
- Manual integration test against an instance of vCenter
**Acceptance Criteria:**
Code and tests implemented.

### PR DESCRIPTION
**Pull request by lcastellano**
Fixes #6504

Implement Role creation/validation/repair for ops-user.

### CODE REVIEW

File: cmd/vic-machine/ops-user/rbac.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+func (am *AuthzManager) DeleteRoles(ctx context.Context) (int, error) {
+	return am.deleteRoles(ctx, getTargetRoles())
+}
+
+func (am *AuthzManager) RoleList(ctx context.Context) (object.AuthorizationRoleList, error) {
+	return am.getRoleList(ctx)
+}
+
+func (am *AuthzManager) createOrRepairRoles(ctx context.Context, targetRoles []types.AuthorizationRole) (int, error) {
```
Comment (hickeng) [inline cmd/vic-machine/ops-user/rbac.go]: We may need to be able to report the missing privileges/roles/permissions in the case where the user has _not_ checked the `Grant Permissions` box. You should consider splitting assembly/identification of missing permissions from the modification - maybe that's a follow up PR after function's complete but will likely reduce duplication.

File: cmd/vic-machine/ops-user/rbac.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+	Role      types.AuthorizationRole
+}
+
+type RBACConfig struct {
+	Resources []Resource
+}
+
+type PermissionList []types.Permission
+
+var rolePrefix = "vch-"
```
Comment (hickeng) [inline cmd/vic-machine/ops-user/rbac.go]: Can we make this `vic-vch-` - while VCH is a very common acronym to us VIC is the name of the product that most people seem to use.

File: cmd/vic-machine/ops-user/rbac.go [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+	err := mgr.UpdateRole(ctx, fRole.RoleId, fRole.Name, fRole.Privilege)
+
+	return true, err
+}
+
+func getTargetRoles() []types.AuthorizationRole {
+	count := len(RBACConf.Resources)
+	roles := make([]types.AuthorizationRole, 0)
+	dSet := make(map[string]bool)
+	for i := 0; i < count; i++ {
```
Comment (hickeng) [inline cmd/vic-machine/ops-user/rbac.go]: why not use `for i := range RBACConf.Resources`?

File: cmd/vic-machine/ops-user/rbac.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+		name := rolePrefix + RBACConf.Resources[i].Role.Name
+		// Discard duplicates
+		if _, found := dSet[name]; !found {
```
Comment (hickeng) [inline cmd/vic-machine/ops-user/rbac.go]: Not sure why you're discarding duplicates given it's an inline struct - should we `panic` instead as it seems like it's programmer error in this circumstance?

File: cmd/vic-machine/ops-user/rbac_test.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+	if err != nil {
+		t.SkipNow()
+	}
+
+	mgr := NewAuthzManager(ctx, session, "ops-user")
+
+	doTestRoles(ctx, t, mgr)
+}
+
+func initRoles(ctx context.Context, t *testing.T, am *AuthzManager,
```
Comment (hickeng) [inline cmd/vic-machine/ops-user/rbac_test.go]: ditto line wrapping comment

File: cmd/vic-machine/ops-user/rbac_test.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+		}
+		if role.Name == targetRoleName2 {
+			removePrivilege(&role, "VirtualMachine.Config.AddNewDisk")
+			am.authzManager.UpdateRole(ctx, role.RoleId, role.Name, role.Privilege)
+		}
+	}
+
+	// Test
+	roleCount = 2
+	count, err = am.createOrRepairRoles(ctx, targetRoles)
```
Comment (hickeng) [inline cmd/vic-machine/ops-user/rbac_test.go]: There are `testify/assert` (continue but fail test) and `testify/require` (fail immediately and stop) packages that simplify this conditional logic and allow for embedding textual commentary in the test output. It makes the tests much more declarative.

File: cmd/vic-machine/ops-user/rbac.go (hunk 1 above)
Comment (lcastellano) [reply cmd/vic-machine/ops-user/rbac.go]: I am thinking of using a different approach for verification of pre-existing permissions. I would log in as the ops-user and use the vmomi API that checks if the user has the permission to do an operation on a resource. I still have to investigate the details of that would work but I doubt there will be much code sharing at the Role level.

File: cmd/vic-machine/ops-user/rbac.go (hunk 4 above)
Comment (lcastellano) [reply cmd/vic-machine/ops-user/rbac.go]: In the inline structure there are roles that legitimately appear in multiple entries. For instance, the Datastore Role applies to Cluster, Datastore and SANDatastore. This method returns the set of roles (no duplication).

File: cmd/vic-machine/ops-user/rbac_test.go (hunk 6 above)
Comment (lcastellano) [reply cmd/vic-machine/ops-user/rbac_test.go]: I left one explicit check. It is:

if checkCount && count != roleCount {
   // fail
}

I tried require.True() but the code becomes less readable in my opinion.

**Pull request merged**